pennylane/templates/subroutines/qpe.py

```diff
@@ -0,0 +1,123 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the ``QuantumPhaseEstimation`` template.
+"""
+import pennylane as qml
+from pennylane.templates.decorator import template
+from pennylane.wires import Wires
+
+
+@template
+def QuantumPhaseEstimation(unitary, target_wires, estimation_wires):
+    r"""Performs the
+    `quantum phase estimation <https://en.wikipedia.org/wiki/Quantum_phase_estimation_algorithm>`__
+    circuit.
+
+    Given a unitary matrix :math:`U`, this template applies the circuit for quantum phase
+    estimation. The unitary is applied to the qubits specified by ``target_wires`` and :math:`n`
+    qubits are used for phase estimation as specified by ``estimation_wires``.
+
+    .. figure:: ../../_static/templates/subroutines/qpe.svg
+        :align: center
+        :width: 60%
+        :target: javascript:void(0);
+
+    This circuit can be used to perform the standard quantum phase estimation algorithm, consisting
+    of the following steps:
+
+    #. Prepare ``target_wires`` in an eigenstate of :math:`U`. If that eigenstate has a
```

Comment:
Quantum phase estimation can be used more generally, starting from a state that is not an eigenstate. Can we rephrase the docstring to saying something like "Prepare ``target_wires`` in a given state. The effect of the algorithm of the algorithm can be analysed by considering it to be an eigenstate of U. Other states can then be written in the eigenbasis of U"

Reply:
Thanks, have tried to mention that other input preparations can be considered.